### PR TITLE
Rock Paper Scissors

### PR DESCRIPTION
**Pull request by ventilooo**
A new task from rosettacode.

### CODE REVIEW

File: examples/rosettacode/rock_paper_scissor.nit [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#author: Istvan szalai 22/05/2015
+
+fun core(player, computer: Int): Int do
+	if player == computer then return 0
+	if player > computer then return 1
+	return 2
+
+end
+
+fun resultMessage(number: Int): String do
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: We don't encourage the use of camelCase in nit, generally variables and functions are expressed using snake_case

File: examples/rosettacode/rock_paper_scissor.nit [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	if number == 0 then return "it's a tie !"
+	if number == 1 then return "You win"
+	return "You loose"
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: :s/loose/lose

File: examples/rosettacode/rock_paper_scissor.nit [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	if std == "scissors" then return 6
+	return 7
+end
+
+fun computer: Int do return 3.rand + 5
+
+var x = ""
+var gameLoop = true
+var test = false
+while gameLoop do
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: Instead of a while true like this, use a `loop end` structure, especially since gameLoop is never used

File: examples/rosettacode/rock_paper_scissor.nit [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	if std == "rock" then return 5
+	if std == "scissors" then return 6
+	return 7
+end
+
+fun computer: Int do return 3.rand + 5
+
+var x = ""
+var gameLoop = true
+var test = false
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: `test` is defined, but never used afterwards

File: examples/rosettacode/rock_paper_scissor.nit [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+	return "You loose"
+end
+
+fun playMessage(number: Int): String do
+	if number == 5 then return "The computer played rock;"
+	if number == 6 then return "the computer played scissors;"
+	return "computer played paper;"
+end
+
+fun convert(std: String): Int do
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: `std` is a strange name for a string, wouldn't you think ? Why don't you name it `cmd` or something else ?

File: examples/rosettacode/rock_paper_scissor.nit [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+# Task: Programming_Task/Rock-paper-scissors
+# SEE: <http://rosettacode.org/wiki/Rock-paper-scissors>
+#
+#author: Istvan szalai 22/05/2015
+
+fun core(player, computer: Int): Int do
+	if player == computer then return 0
+	if player > computer then return 1
+	return 2
+
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: Why the empty line ?

File: examples/rosettacode/rock_paper_scissor.nit [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+fun resultMessage(number: Int): String do
+	if number == 0 then return "it's a tie !"
+	if number == 1 then return "You win"
+	return "You loose"
+end
+
+fun playMessage(number: Int): String do
+	if number == 5 then return "The computer played rock;"
+	if number == 6 then return "the computer played scissors;"
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: Try to be consistent in the way you phrase messages, some start with an upper-case, some don't, some have a `The` some others don't, etc.

**Comment by R4PaSs** [comment]
You forgot to add a commit message, please change that before we merge your PR.

And aside from the remarks, :shipit:

File: examples/rosettacode/rock_paper_scissor.nit [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+while gameLoop do
+	print "Choose your weapon|exit"
+	x = gets
+	if x == "exit" then break
+	x.to_i
+	var z = computer
+	print playMessage(z)
+	print resultMessage(core(convert(x), z))
+end
+
```
Comment (Tagachi) [inline examples/rosettacode/rock_paper_scissor.nit]: Blank line

File: examples/rosettacode/rock_paper_scissor.nit [hunk 9]
```diff
@@ -0,0 +1,48 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+# This program is public domain
+#
+# Task: Programming_Task/Rock-paper-scissors
+# SEE: <http://rosettacode.org/wiki/Rock-paper-scissors>
+#
+#author: Istvan szalai 22/05/2015
+
+fun core(player, computer: Int): Int do
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: Try to comment your methods too, this one looks strange (and is never used, why is it here ?)

File: examples/rosettacode/rock_paper_scissor.nit [hunk 10]
```diff
@@ -0,0 +1,48 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+# This program is public domain
+#
+# Task: Programming_Task/Rock-paper-scissors
+# SEE: <http://rosettacode.org/wiki/Rock-paper-scissors>
+#
+#author: Istvan szalai 22/05/2015
```
Comment (Tagachi) [inline examples/rosettacode/rock_paper_scissor.nit]: Whitespace ```# author:...```

File: examples/rosettacode/rock_paper_scissor.nit [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+#return a random value for the computer player input
+fun computer_player: Int do return 3.rand + 5 end
+
+#Game loop
+var x = ""
+loop do
+	print "Choose your weapon|exit"
+	x = gets
+	if x == "exit" then break
+	x.to_i
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: This doesn't do anything.

File: examples/rosettacode/rock_paper_scissor.nit [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	if number == 5 then return "The computer played rock;"
+	if number == 6 then return "The computer played scissors;"
+	return "The computer played paper;"
+end
+
+#return the converted string for the core function
+fun convert(str: String): Int do
+	if str == "rock" then return 5
+	if str == "scissors" then return 6
+	return 7
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: If the player writes "asdf" there should be an error instead of playing paper.

File: examples/rosettacode/rock_paper_scissor.nit [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+# Task: Programming_Task/Rock-paper-scissors
+# SEE: <http://rosettacode.org/wiki/Rock-paper-scissors>
+#
+# author: Istvan szalai 22/05/2015
+
+#return a value to dertermine who win the match
+fun win_or_lose(player, computer: Int): Int do
+	if player == computer then return 0
+	if player > computer then return 1
+	return 2
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: This logic is flawed, paper (7) beats scissors (6), and scissors beat rock.

File: examples/rosettacode/rock_paper_scissor.nit [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+end
+
+#return a string with about the match winner or looser
+fun result_of_the_match_message(number: Int): String do
+	if number == 0 then return "It's a tie !"
+	if number == 1 then return "You win"
+	return "You lose"
+end
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: You should unify the last 2 methods. `win_or_lose` can directly print the result as a string. There would be less magic numbers!

**Comment by xymus** [comment]
To hide all those magic numbers, normally we would encapsulate them in a class. But for such a small script you could use simple top-level methods:

~~~
fun rock: Int do return 5
fun scissor: Int do return 6
fun paper: Int do return 7
...

if gets == "rock" then return rock
~~~

**Comment by xymus** [comment]
ok to test

File: examples/rosettacode/rock_paper_scissor.nit [hunk 15]
```diff
@@ -0,0 +1,50 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+# This program is public domain
+#
+# Task: Programming_Task/Rock-paper-scissors
+# SEE: <http://rosettacode.org/wiki/Rock-paper-scissors>
+#
+# author: Istvan szalai 22/05/2015
+
+#return a value to dertermine who win the match
```
Comment (Morriar) [inline examples/rosettacode/rock_paper_scissor.nit]: Comments on functions should be formatted this way:

* There is a space between the `#` and the start of the comment
* There is an uppercase at the start of the line
* There is (optionally) a dot at the end of the line

Example:

~~~nit
# This is a nice and tidy Nit commend.
~~~

Comment (Morriar) [inline examples/rosettacode/rock_paper_scissor.nit]: Furthermore, this comment should explain your logic and why you return 0, 1 or 2.

File: examples/rosettacode/rock_paper_scissor.nit [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+fun win_or_lose(player, computer: Int): Int do
+	if player == computer then return 0
+	if player > computer then return 1
+	return 2
+end
+
+#return a string with about the match winner or looser
+fun result_of_the_match_message(number: Int): String do
+	if number == 0 then return "It's a tie !"
```
Comment (Morriar) [inline examples/rosettacode/rock_paper_scissor.nit]: In English, there is no space between the text and the exclamation mark.

File: examples/rosettacode/rock_paper_scissor.nit [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+	while not test do
+		print "Choose your weapon|exit"
+		x = gets
+		if x == "exit" then break
+		if x == "paper" or x == "rock" or x == "scissors" then test = true
+		if test == false then print"Unknow input!"
+	end
+	test = false
+	var z = computer_player
+	print computer_message(z)
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: This is much more complicated than it needs to be. The 2 previous calls and methods can be replaced by:
~~~
var z = ["rock", "paper", "scissors"].rand
print "The computer played {z}"
~~~

Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: You can the remove `computer_player, convert and computer message`.

File: examples/rosettacode/rock_paper_scissor.nit [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+# Game loop.
+var x = ""
+var test = false
+loop do
+	while not test do
+		print "Choose your weapon|exit"
+		x = gets
+		if x == "exit" then break
+		if x == "paper" or x == "rock" or x == "scissors" then test = true
+		if test == false then print"Unknow input!"
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: *`print "Unknown input"`

You should tell the player then choices in the prompt or the error message.

**Comment by ventilooo** [comment]
Waiting your comment @Morriar  @R4PaSs  @xymus

File: examples/rosettacode/rock_paper_scissor.nit [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+fun computer_message(number: String): String do
+	if number == "rock" then return "The computer played rock"
+	if number == "scissors" then return "The computer played scissors"
+	return "The computer played paper"
+end
+
+# Return a random value for the computer player input.
+fun computer_player : String do
+	var tab = ["rock", "scissors", "paper"]
+	return tab.rand
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: You could directly return rand on the literal array, the variable is not really useful here

File: examples/rosettacode/rock_paper_scissor.nit [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+# Game loop.
+var x = ""
+var test = false
+loop
+	while not test do
+		print "Choose your weapon|exit"
+		x = gets
+		if x == "exit" then break
+		if x == "paper" or x == "rock" or x == "scissors" then test = true
+		if test == false then print"Invalid choice (rock/paper/scissors"
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: `if not test` ?

Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: And please add a space between the print and the String

Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: Also, close the parenthesis in the string

File: examples/rosettacode/rock_paper_scissor.nit [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+# Game loop.
+var x = ""
+var test = false
+loop
+	while not test do
+		print "Choose your weapon|exit"
+		x = gets
+		if x == "exit" then break
+		if x == "paper" or x == "rock" or x == "scissors" then test = true
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: These tests could be replaced by a simple `["paper", "rock", "scissors"].has(x)`

Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: And both tests could be merged then

File: examples/rosettacode/rock_paper_scissor.nit [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+end
+
+# Game loop.
+var x = ""
+var test = false
+loop
+	while not test do
+		print "Choose your weapon|exit"
+		x = gets
+		if x == "exit" then break
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: I'm a bit concerned with this `break`, I think it'll just leave the while, not the encompassing loop, you should probably do a `break LABEL` if this is what you meant here

File: examples/rosettacode/rock_paper_scissor.nit [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+	if player == computer then return 2
+	if player == "rock" and computer == "scissors" then return 1
+	if player == "rock" and computer == "paper" then return 0
+	if player == "paper" and computer == "rock" then return 1
+	if player == "paper" and computer == "scissors" then return 0
+	if player == "scissors" and computer == "paper" then return 1
+	return 0
+end
+
+# Return a string with about the match winner or looser.
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: `with about` ? This does not make sense.

You should probably just say that the method returns whether the match gives a winner or a loser.

Or better then, since the number is exclusively 0, 1 or 2, you could have a literal array access in your main on the right string, hence this method could disappear altogether

File: examples/rosettacode/rock_paper_scissor.nit (hunk 21 above)
Comment (privat) [inline examples/rosettacode/rock_paper_scissor.nit]: I do not recommend using `["paper", "rock", "scissors"].has(x)` since a new Array will be allocated each time. Here this is not an issue but since it is in `examples`, it is not something I want to encourage as a good way to do things.

Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: I'm a bit surprised to hear that ! I thought that the array was
instantiated once only, but I guess it makes sense since arrays are indeed
mutable.

Maybe we should proceed differently then ? Have the literal arrays
immutable ? If we go with a literal array, most of the time it is for
read-only purposes, in this sense it could be varonced ?

2015-06-03 10:49 GMT-04:00 Jean Privat <notifications@github.com>:

File: examples/rosettacode/rock_paper_scissor.nit (hunk 19 above)
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: Come to think of it, for the two lines it represents, it could be in the main I think, this function could diappear too

File: examples/rosettacode/rock_paper_scissor.nit (hunk 21 above)
Comment (privat) [inline examples/rosettacode/rock_paper_scissor.nit]: Unfortunately, immutability will not helps you here because there is two issues: (i) what bad things could happen if I reuse the array but it has escaped? Immutability will helps a little because it ensure that the escaped array win not change (but the element may change, you need deep immutability in fact). (ii) if we reuse the array, how confident are we that the initial data make sense in the current context.

To illustrate (ii) lets use this example

~~~nit
fun is_one_or_the_other(object, one, other: Object): Bool do
  return [one, other].has(object)
end

print is_one_or_the_other("a", "a", "c") # true
print is_one_or_the_other("a", "b", "c") # false
~~~

If you assume that the literal array is reused then this means that its content ["a", "c"] is also reused, thus the second call of the method will not have the expected result.

To summarize, in order to reuse a literal array you need that
 * the array is deep-immutable
 * the value of the elements are independent of the state of the computer (parameters, `sys`, I/O, or anything that may depend on `sys`, I/O, or other things)
 
Both points will be very hard to make them fit with the Nit philosophy, so unlikely to be specific or implemented.

Note that if you can control the mutation, the escaping and the refinements, then you can use the infamous and undocumented `once` to reuse the literal array if you want.

Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: You're right, I was thinking a bit too fast in this case.

Well then, a variable is good enough in this case I guess and ok, literal arrays won't be cached anytime soon

File: examples/rosettacode/rock_paper_scissor.nit [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+# Return a random value for the computer player input.
+fun computer_player : String do
+	var tab = ["rock", "scissors", "paper"]
+	return tab.rand
+end
+
+# Game loop.
+var player_input = ""
+var tab = ["rock", "scissors", "paper"]
+var tab_message = ["You lose", "You win!", "It's a tie!"]
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: Why is "You lose" the only one without a ! ?

File: examples/rosettacode/rock_paper_scissor.nit [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+loop
+	loop
+		print "Choose your weapon|exit"
+		player_input = gets
+		if player_input == "exit" then exit 0
+		if tab.has(player_input) then break
+		if not tab.has(player_input) then print "(Invalid choice (rock/paper/scissors)"
```
Comment (R4PaSs) [inline examples/rosettacode/rock_paper_scissor.nit]: Why is there a parenthesis before `Invalid choice` ?

File: examples/rosettacode/rock_paper_scissor.nit [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+# Game loop.
+var index = 0
+var player_input = ""
+var tab = ["rock", "scissors", "paper"]
+var tab_message = ["You lose", "You win!", "It's a tie!"]
+var tab_computer_message = ["The computer played rock", "The computer played scissors", "The computer played paper"]
+loop
+	loop
+		print "Choose your weapon|exit"
+		player_input = gets
```
Comment (privat) [inline examples/rosettacode/rock_paper_scissor.nit]: I suggest that you exit the program (forfeit?) if end of file.

~~~nit
if stdin.eof then return
~~~

File: examples/rosettacode/rock_paper_scissor.nit [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+		if player_input == "exit" then exit 0
+		if tab.has(player_input) then break
+		if not tab.has(player_input) then print "Invalid choice (rock/paper/scissors)"
+	end
+	var computer_input = tab.rand
+		for i in [0..tab.length[ do	
+			if computer_input == tab[i] then  index = i 
+		end
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: The loop could simply be: `for i in tab do`

And there's an extra tabulation on this block.

File: examples/rosettacode/rock_paper_scissor.nit [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+	print tab_computer_message[index]
+	print tab_message[win_or_lose(player_input,computer_input)]
+	if stdin.eof then return
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: I'm pretty sure this must be within the inner loop, where @privat left his comment.

File: examples/rosettacode/rock_paper_scissor.nit [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+		player_input = gets
+		if player_input == "exit" then exit 0
+		if tab.has(player_input) then break
+		if not tab.has(player_input) then print "Invalid choice (rock/paper/scissors)"
+	end
+	var computer_input = tab.rand
+		for i in [0..tab.length[ do	
+			if computer_input == tab[i] then  index = i 
+		end
+	print tab_computer_message[index]
```
Comment (xymus) [inline examples/rosettacode/rock_paper_scissor.nit]: Actually, I see what you did here... Why not just use `print "The computer played {computer_input}"` instead of the 4 previous lines?

**Comment by xymus** [comment]
It is much cleaner, we a getting closer to a merge!
